sys/kern/devfs.c

```diff
@@ -75,142 +95,175 @@ static devfs_node_t *devfs_find_child(devfs_node_t *parent,
-  dn->dn_uid = 0;
-  dn->dn_gid = 0;
-  dn->dn_nlinks = 1;
+  devfs_node_t *dn = devfs_node_create(name, mode);
   dn->dn_parent = parent;
-  dn->dn_name = kstrndup(M_STR, name, DEVFS_NAME_MAX);
   TAILQ_INSERT_TAIL(&parent->dn_children, dn, dn_link);
-  parent->dn_nlinks++;
+  if (mode & S_IFDIR)
+    parent->dn_nlinks++;
   *dnp = dn;
+
   return 0;
 }
 
-int devfs_unlink(devfs_node_t *dn) {
-  SCOPED_MTX_LOCK(&devfs.lock);
+/* FIXME should be defined as static */
+void devfs_free(devfs_node_t *dn) {
+  assert(dn->dn_vnode->v_usecnt == 0);
+  kfree(M_DEVFS, dn);
+}
 
-  assert(dn->dn_parent != NULL);
+/*
+ * Fileops interface for device nodes.
+ */
 
-  /* Only allow removal of empty directories. */
-  if (dn->dn_vnode->v_type == V_DIR && !TAILQ_EMPTY(&dn->dn_children))
-    return ENOTEMPTY;
+static int devfs_fop_read(file_t *fp, uio_t *uio) {
+  devnode_t *dev = fp->f_data;
+  return dev->ops->d_read(dev, uio);
+}
 
-  TAILQ_REMOVE(&dn->dn_parent->dn_children, dn, dn_link);
-  vnode_drop(dn->dn_vnode);
-  return 0;
+static int devfs_fop_write(file_t *fp, uio_t *uio) {
+  devnode_t *dev = fp->f_data;
+  return dev->ops->d_write(dev, uio);
 }
 
-void devfs_free(devfs_node_t *dn) {
-  assert(dn->dn_vnode->v_usecnt == 0);
-  kfree(M_STR, dn->dn_name);
-  kfree(M_DEVFS, dn);
+static int devfs_fop_close(file_t *fp) {
+  devnode_t *dev = fp->f_data;
+  refcnt_release(&dev->refcnt);
+  return dev->ops->d_close(dev, fp->f_flags);
 }
 
-/* Free a devfs directory after unlinking it. */
-static int devfs_vop_reclaim(vnode_t *v) {
-  assert(v->v_type == V_DIR);
-  devfs_free(vn2dn(v));
+static int devfs_fop_seek(file_t *fp, off_t offset, int whence,
+                          off_t *newoffp) {
+  devnode_t *dev = fp->f_data;
+  bool seekable = dev->ops->d_type & DT_SEEKABLE;
+
+  if (!seekable)
+    return ESPIPE;
+
+  if (whence == SEEK_CUR) {
+    offset += fp->f_offset;
+  } else if (whence == SEEK_END) {
+    offset += dev->size;
+  } else if (whence != SEEK_SET) {
+    return EINVAL;
+  }
+
+  if (offset < 0) {
+    offset = 0;
+  } else if ((size_t)offset > dev->size) {
+    offset = dev->size;
+  }
+
+  *newoffp = fp->f_offset = offset;
   return 0;
 }
 
+static int devfs_fop_ioctl(file_t *fp, u_long cmd, void *data) {
+  devnode_t *dev = fp->f_data;
+  return dev->ops->d_ioctl(dev, cmd, data, fp->f_flags);
+}
+
+static fileops_t devfs_fileops = {
+  .fo_read = devfs_fop_read,
+  .fo_write = devfs_fop_write,
+  .fo_close = devfs_fop_close,
+  .fo_seek = devfs_fop_seek,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = devfs_fop_ioctl,
+};
+
+/*
+ * Devfs device v-node operations.
+ */
+
 static int devfs_vop_getattr(vnode_t *v, vattr_t *va) {
-  devfs_node_t *dn = vn2dn(v);
-  memset(va, 0, sizeof(vattr_t));
+  devfs_node_t *dn = devfs_node_of(v);
+
+  bzero(va, sizeof(vattr_t));
   va->va_mode = dn->dn_mode;
   va->va_uid = dn->dn_uid;
   va->va_gid = dn->dn_gid;
+  va->va_size = 0; /* only opened device nodes have size */
```

Comment:
This is very unclear to me. Is this function meant to be called only when the node is not opened? Why is that so? If that's something that's not totally obvious given the semantics, I think there should be an assert for that.

Reply:
Exactly. Consider a storage device with exchangeable media (floppy, cd/dvd). Unless you open a device, which will succeed only if the media is present in drive, you don't know its size.